senderSampleApp/src/main/java/org/kinecosystem/sampleSenderApp/kindiscover/SendKinService.java

```diff
@@ -2,50 +2,52 @@
 import org.kinecosystem.sampleSenderApp.SenderApplication;
 
 import java.math.BigDecimal;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
 
 import kin.sdk.Transaction;
 import kin.sdk.TransactionId;
 
 public class SendKinService extends SendKinServiceBase {
+    private ExecutorService executorService = Executors.newCachedThreadPool();
 
     @Override
-    public @NonNull
-    KinTransferComplete transferKin(@NonNull String toAddress, int amount, @NonNull String memo) throws KinTransferException {
+    public
+    void transferKin(@NonNull String toAddress, int amount, @NonNull String memo, @NotNull KinTransferCallback callback) {
         SampleWallet sampleWallet = ((SenderApplication) getApplicationContext()).getSampleWallet();
-        String sourceAddress = "None";
-
-        if (!sampleWallet.hasActiveAccount()) {
-
-            throw new KinTransferException(sourceAddress, "Cannot transfer Kin. Account not initialized");
-        }
-
-        try {
-            sourceAddress = sampleWallet.getAccount().getPublicAddress();
 
-            int fee = 100; // no whitelisting for sample app, so using a fee
-            Transaction transaction = sampleWallet.getAccount().buildTransactionSync(toAddress,
-                    new BigDecimal(amount), fee, memo);
-            TransactionId transactionId = sampleWallet.getAccount().sendTransactionSync(transaction);
-
-            // here you may add some code to add the transaction details to
-            // your app's transaction history metadata
-
-            return new KinTransferComplete(sourceAddress, transactionId.id(), transaction.getMemo());
-
-        } catch (Exception e) {
-
-            e.printStackTrace();
-
-            throw new KinTransferException(sourceAddress,
-                    "Cannot transfer Kin. Exception " + e + ", with message " + e.getMessage());
+        transferKinCallback = callback;
 
+        if (!sampleWallet.hasActiveAccount()) {
+            if (transferKinCallback != null)
+                transferKinCallback.onError(new KinTransferException("None", "Cannot transfer Kin. Account not initialized"));
         }
+        executorService.execute(() -> {
+            try {
+                String sourceAddress = sampleWallet.getAccount().getPublicAddress();
+                int fee = 100; // no whitelisting for sample app, so using a fee
+                Transaction transaction = sampleWallet.getAccount().buildTransactionSync(toAddress,
+                        new BigDecimal(amount), fee, memo);
+                TransactionId transactionId = sampleWallet.getAccount().sendTransactionSync(transaction);
+                // here you may add some code to add the transaction details to
+                // your app's transaction history metadata
+                if (transferKinCallback != null)
+                    transferKinCallback.onSuccess(new KinTransferComplete(sourceAddress, transactionId.id(), transaction.getMemo()));
```

Comment:
you should callback should be called on the UI thread

Reply:
Actually, since we agreed to make this backward compatible - it means that we can leave the sample app as is.